### PR TITLE
feat: Infrastructure Manager workspace blueprint

### PR DESCRIPTION
**Pull request by josephdt12**
This blueprint creates an opinionated workflow for actuating Terraform resources using Infrastructure Manager using GitHub/GitLab and Cloud Build.

### CODE REVIEW

File: test/integration/im_cloudbuild_workspace_github/im_cloudbuild_workspace_github_test.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	updatedState := "closed"
+	pr.State = &updatedState
+	_, _, err := gh.client.PullRequests.Edit(ctx, gh.owner, gh.repo, *pr.Number, pr)
+	if err != nil {
+		gh.t.Fatal(err.Error())
+	}
+}
+
+func TestIMCloudBuildWorkspaceGitHub(t *testing.T) {
+	var ctx = context.Background()
```
Comment (bharathkkb) [inline test/integration/im_cloudbuild_workspace_github/im_cloudbuild_workspace_github_test.go:97]:
```suggestion
	ctx := context.Background()
```

File: build/int.cloudbuild.yaml [hunk 2]
```diff
@@ -26,16 +26,22 @@ steps:
...
+  - 'TF_VAR_im_github_pat_secret=$_IM_GITHUB_PAT_SECRET_ID'
+  secretEnv: ['IM_GITHUB_PAT']
 - id: init-all
   waitFor:
     - prepare
   name: 'gcr.io/cloud-foundation-cicd/$_DOCKER_IMAGE_DEVELOPER_TOOLS:$_DOCKER_TAG_VERSION_DEVELOPER_TOOLS'
   args: ['/bin/bash', '-c', 'cft test run all --stage init --verbose']
+  secretEnv: ['IM_GITHUB_PAT']
+  env:
+  - 'TF_VAR_im_github_pat_secret=$_IM_GITHUB_PAT_SECRET_ID'
```
Comment (bharathkkb) [inline build/int.cloudbuild.yaml:38]: nit: I think this is unnecessary

**Review by bharathkkb — APPROVED** [review]
Thanks for the PR @josephdt12

File: examples/im_cloudbuild_workspace_github/main.tf [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  project_id    = var.project_id
+  deployment_id = "im-example-github-deployment"
+
+  tf_repo_type           = "GITHUB"
+  im_deployment_repo_uri = "https://github.com/im-goose/infra-manager-git-example.git"
+  im_deployment_ref      = "main"
+  im_tf_variables        = "project_id=${var.project_id}"
+  infra_manager_sa_roles = ["roles/compute.networkAdmin"]
+
+  github_app_installation_id   = "47590865"
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply examples/im_cloudbuild_workspace_github/main.tf:30]: Added comment and linked to the Cloud Build documentation on it as well.

File: build/int.cloudbuild.yaml (hunk 2 above)
Comment (josephdt12) [reply build/int.cloudbuild.yaml:38]: Currently following the [example here](https://github.com/GoogleCloudPlatform/terraform-google-tf-cloud-agents/blob/main/build/int.cloudbuild.yaml#L184-L187) which sets up a similar secret to be used here [https://github.com/GoogleCloudPlatform/terraform-google-tf-cloud-agents/blob/main/build/int.cloudbuild.yaml#L156](https://github.com/GoogleCloudPlatform/terraform-google-tf-cloud-agents/blob/main/build/int.cloudbuild.yaml#L156).

Maybe the other way of explicitly creating the TF_VAR link could work, just involves some other changes I don't recall at the top of my head.

File: test/integration/im_cloudbuild_workspace_github/im_cloudbuild_workspace_github_test.go (hunk 1 above)
Comment (josephdt12) [reply test/integration/im_cloudbuild_workspace_github/im_cloudbuild_workspace_github_test.go:97]: Woops, updated.

File: modules/im_cloudbuild_workspace/versions.tf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      version = ">= 4.17, != 4.31.0, < 6"
+    }
+    random = {
+      source  = "hashicorp/random"
+      version = ">= 3.6.0"
+    }
+  }
+
+  provider_meta "google" {
+    module_name = "blueprints/terraform/terraform-google-bootstrap:tf_cloudbuild_workspace/v7.0.0"
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/versions.tf:38]: Updated, although is there any issues with any versioning here given this is unreleased at the moment?

File: modules/im_cloudbuild_workspace/versions.tf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  required_version = ">= 1.5.7"
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/versions.tf:18]: Updated to just the minimum IM version of 1.2.3.

File: modules/im_cloudbuild_workspace/templates/create-preview.sh.tftpl [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+echo "Checking if deployment ${deployment_id} already exists"
+gcloud infra-manager deployments describe projects/${project_id}/locations/${location}/deployments/${deployment_id} --no-user-output-enabled
+DEPLOYMENT_EXISTS=$(echo $?)
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/templates/create-preview.sh.tftpl:18]: Yeah, that makes more sense. Added to filter on the deployment ID and pipe it into a `wc -l`.

File: modules/im_cloudbuild_workspace/sa.tf [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  create_cloudbuild_sa    = var.cloudbuild_sa == ""
+  cloudbuild_sa           = local.create_cloudbuild_sa ? google_service_account.cb_sa[0].id : var.cloudbuild_sa
+  cloudbuild_sa_email     = element(split("/", local.cloudbuild_sa), length(split("/", local.cloudbuild_sa)) - 1)
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/sa.tf:20]: Added comment.

File: modules/im_cloudbuild_workspace/repo.tf [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+data "google_project" "project" {
+  project_id = var.project_id
+}
+
+data "google_iam_policy" "serviceagent_secretAccessor" {
+  binding {
+    role    = "roles/secretmanager.secretAccessor"
+    members = ["serviceAccount:service-${data.google_project.project.number}@gcp-sa-cloudbuild.iam.gserviceaccount.com"]
+  }
+}
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/repo.tf:37]: Makes sense, moved it over. Was more of a carry over from CB's documentation.

File: modules/im_cloudbuild_workspace/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+output "trigger_location" {
+  description = "Location of for Cloud Build triggers created in the workspace. Matches `location` if not given."
+  value       = var.trigger_location
+}
+
+output "im_deployment_repo_uri" {
+  description = "The URI of the repo where the Terraform configs are stored and triggers are created for"
+  value       = var.im_deployment_repo_uri
+}
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/outputs.tf:30]: That's valid, I was using it for the integration test, but I can have those values hard-coded to match the example.

File: modules/im_cloudbuild_workspace/github.tf [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// Personal access token from VCS.
+resource "google_secret_manager_secret_version" "github_token_secret_version" {
+  count       = local.create_github_secret ? 1 : 0
+  secret      = google_secret_manager_secret.github_token_secret[0].id
+  secret_data = var.github_personal_access_token
+}
+
+resource "google_secret_manager_secret_iam_policy" "github_iam_policy" {
```
[the remark this answers is not in the document]
Comment (josephdt12) [reply modules/im_cloudbuild_workspace/github.tf:50]: Updated to that one instead.

File: modules/im_cloudbuild_workspace/versions.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/im_cloudbuild_workspace/versions.tf]: The versioning should be okay, our release tooling should bump the version string in next release PR

File: build/int.cloudbuild.yaml (hunk 2 above)
Comment (bharathkkb) [reply build/int.cloudbuild.yaml]: In this case it was just a nit that this init step wont care about the env var so we could skip it

**Pull request merged**